### PR TITLE
Introduce AAD auth for android sdk

### PR DESCRIPTION
**Pull request by manuchandel**
## Introduction

The app center SDK will now allow authentication via AAD.
Some conditions have been enforced in Auth.java 
Configuration specs for MSAL can be read [here](https://github.com/AzureAD/microsoft-authentication-library-for-android/wiki/Configuring-your-app)

### Config File

```
{
  "client_id" : "<CLIENT_ID_FROM_https://apps.dev.microsoft.com>",
  "authorization_user_agent" : "DEFAULT",
  "redirect_uri" : "<CLIENT_ID_FROM_https://apps.dev.microsoft.com>://auth",
  "authorities" : [
    {
      "type": "AAD",
      "audience": {
       "type": "AzureADandPersonalMicrosoftAccount"
      }
    }
  ]
}
```

### AAD conditions

1. Should have audience type.
2. Should not have authority url.

### B2C conditions

1. Should have authority URL.
2. Should not have audience type.

## code changes

1. Introduced constants for AAD auth type.
1. Added logic to enforce the conditions mentioned above.
1. Changed error messages for new configuration.

Please have a look at our [guidelines for contributions](https://github.com/microsoft/appcenter-sdk-android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

A few sentences describing the overall goals of the pull request.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 1]
```diff
@@ -461,24 +469,58 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
             for (int i = 0; i < authorities.length(); i++) {
                 JSONObject authority = authorities.getJSONObject(i);
                 if (authority.optBoolean(AUTHORITY_DEFAULT) && AUTHORITY_TYPE_B2C.equals(authority.getString(AUTHORITY_TYPE))) {
-                    authorityUrl = authority.getString(AUTHORITY_URL);
-                    break;
+                    type = AUTHORITY_TYPE_B2C;
+                } else if (authority.optBoolean(AUTHORITY_DEFAULT) && AUTHORITY_TYPE_AAD.equals(authority.getString(AUTHORITY_TYPE))) {
+                    type = AUTHORITY_TYPE_AAD;
                 }
+
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]:
```suggestion
```

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 2]
```diff
@@ -461,24 +469,58 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
+                JSONObject audience = authority.optJSONObject(Audience);
+                if (audience != null) {
+                    audienceType = audience.getString(Audience_Type);
+                }
+
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]:
```suggestion
```

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java (hunk 1 above)
Comment (guperrot) [reply sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: Our coding conventions is not to have blank lines unless there is a comment.

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 3]
```diff
@@ -461,24 +469,58 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
+                authorityUrl = authority.optString(AUTHORITY_URL);
             }
-            if (authorityUrl != null) {
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]:
```suggestion
```

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 4]
```diff
@@ -461,24 +469,58 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
+                        && !audienceType.equals(Audience_Type_AzureAdMultipleOrgs)
+                        && !audienceType.equals(Audience_Type_AzureAdAndPersonalMicrosoftAccount)
+                        && !audienceType.equals(Audience_Type_PersonalMicrosoftAccount)) {
+                    throw new IllegalStateException(String.format("AAD authority is configured incorrectly. Audience type=%s is unknown.", audienceType));
+                }
             } else {
-                throw new IllegalStateException("Cannot find a b2c authority configured to be the default.");
+                throw new IllegalStateException("Cannot find a b2c or aad authority configured to be the default.");
             }
+
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]:
```suggestion
```

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 5]
```diff
@@ -461,24 +469,58 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
+            mAuthenticationClient = new PublicClientApplication(mContext, getConfigFile());
+            mAuthorityUrl = authorityUrl;
+            mIdentityScope = identityScope;
+            AppCenterLog.info(LOG_TAG, "Auth service configured successfully.");
+
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]:
```suggestion
```

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Constants.java [hunk 6]
```diff
@@ -86,4 +91,39 @@
      * JSON configuration key for identity scope.
      */
     static final String IDENTITY_SCOPE = "identity_scope";
+
+    /**
+     * JSON configuration key for audience within {@link #AUTHORITIES} array.
+     */
+    static final String Audience = "audience";
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Constants.java]: All constants need to be UPPER_UNDERSCORE_CASING

**Comment by guperrot** [comment]
This PR needs new unit tests to test every new cases.

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 7]
```diff
@@ -461,24 +469,58 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
-            if (authorityUrl != null) {
 
-                /* The remaining validation is done by the library. */
-                mAuthenticationClient = new PublicClientApplication(mContext, getConfigFile());
-                mAuthorityUrl = authorityUrl;
-                mIdentityScope = identityScope;
-                AppCenterLog.info(LOG_TAG, "Auth service configured successfully.");
+            if (type.equals(AUTHORITY_TYPE_B2C)) {
+                if (authorityUrl == null) {
+                    throw new IllegalStateException("B2C authority is configured incorrectly. Authority url is mandatory.");
```
Comment (micarls) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: I would argue checks like this are overzealous. I'm assuming MSAL already throws relevant errors in these cases.

Clement mentioned we need 100% code coverage, which if I understand correctly means that each of these failure lines will require a separate unit test to test it, which feels like a lot of code, and all these checks and all those unit tests seem like more code than they are worth for what it's doing.

Comment (guperrot) [reply sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: Yes the big benefit of coverage forces us to use less if and only ones that actually matter and avoid dead code :)

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Constants.java (hunk 6 above)
Comment (manuchandel) [reply sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Constants.java]: Got it. Changed all constants to upper case.

Comment (micarls) [reply sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Constants.java]: I think you missed Audience itself.

Comment (manuchandel) [reply sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Constants.java]: Thanks Michael

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java (hunk 7 above)
Comment (manuchandel) [reply sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: I added unit tests for this change.

**Comment by manuchandel** [comment]
> Let's remove/simplify all the checks that will be done by MSAL automatically. What the SDK needs to check is only that we can extract some fields we use from JSON in addition to using the msal client.

Made logic simpler

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 8]
```diff
@@ -461,24 +462,26 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
             String authorityUrl = null;
+            String type = null;
             JSONArray authorities = configuration.getJSONArray(AUTHORITIES);
             for (int i = 0; i < authorities.length(); i++) {
                 JSONObject authority = authorities.getJSONObject(i);
                 if (authority.optBoolean(AUTHORITY_DEFAULT) && AUTHORITY_TYPE_B2C.equals(authority.getString(AUTHORITY_TYPE))) {
+                    type = AUTHORITY_TYPE_B2C;
                     authorityUrl = authority.getString(AUTHORITY_URL);
-                    break;
+                } else if (authority.optBoolean(AUTHORITY_DEFAULT) && AUTHORITY_TYPE_AAD.equals(authority.getString(AUTHORITY_TYPE))) {
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: Could factor the check for default even if that increase indentation a little bit.

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 9]
```diff
@@ -461,24 +462,26 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
-                throw new IllegalStateException("Cannot find a b2c authority configured to be the default.");
+            if (type == null) {
+                throw new IllegalStateException("Cannot find a default b2c or aad authority configured to be the default.");
+            } else if (type.equals(AUTHORITY_TYPE_B2C) && authorityUrl == null) {
+                throw new IllegalStateException("Cannot find authority url for b2c.");
             }
+            mAuthenticationClient = new PublicClientApplication(mContext, getConfigFile());
+            mAuthorityUrl = authorityUrl;
+            mIdentityScope = identityScope;
+            AppCenterLog.info(LOG_TAG, "Auth service configured successfully.");
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: Ok code sounds good now.

File: sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java [hunk 10]
```diff
@@ -284,17 +289,41 @@ private void testInvalidConfig(JSONObject jsonConfig) throws Exception {
...
+
+        /* mock public client application for msal*/
+        mockMsalPublicClientApplication();
+
+        /* When we get a payload valid for AppCenter fields but invalid for msal ones. */
+        mockSuccessfulHttpCall(jsonConfig, mHttpClient);
+
+        /* We saved after we downloaded the file. */
         verifyStatic();
         FileManager.write(any(File.class), anyString());
```
Comment (guperrot) [inline sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java]: Let's add verify that we never delete the file when valid

File: sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java [hunk 11]
```diff
@@ -284,17 +289,41 @@ private void testInvalidConfig(JSONObject jsonConfig) throws Exception {
...
         mockSuccessfulHttpCall(jsonConfig, mHttpClient);
 
-        /* We didn't attempt to even save. */
+        /* We saved after we downloaded the file. */
+        verifyStatic();
+        FileManager.write(any(File.class), anyString());
+
+        /* We cleared the cache since the file was invalid. */
+        verifyStatic();
+        FileManager.delete(any(File.class));
```
Comment (guperrot) [inline sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java]: We return a valid msal public client app, so why do we delete the file given we don't simulate an exception? There might be something failing in the test for another reason.

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java [hunk 12]
```diff
@@ -461,24 +462,28 @@ private synchronized void initAuthenticationClient(String configurationPayload)
...
-                /* The remaining validation is done by the library. */
-                mAuthenticationClient = new PublicClientApplication(mContext, getConfigFile());
-                mAuthorityUrl = authorityUrl;
-                mIdentityScope = identityScope;
-                AppCenterLog.info(LOG_TAG, "Auth service configured successfully.");
-            } else {
-                throw new IllegalStateException("Cannot find a b2c authority configured to be the default.");
+            if (type == null) {
+                throw new IllegalStateException("Cannot find a default b2c or aad authority configured to be the default.");
+            } else if (type.equals(AUTHORITY_TYPE_B2C) && authorityUrl == null) {
```
Comment (guperrot) [inline sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: Screen Shot 2019-09-09 at 3 10 42 PM

Android Studio is right, `authority.getString(AUTHORIY_URL)` throws an exception if null.

But in `AuthTest` JSON is mocked. So the setup for `downloadConfigurationWithoutUrl` is wrong.
Accept Android Studio fix and change `downloadConfigurationWithoutUrl` to do

`when(authority.getString("url")).thenThrow(new JSONException("missing"))`

File: sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java [hunk 13]
```diff
@@ -284,17 +289,41 @@ private void testInvalidConfig(JSONObject jsonConfig) throws Exception {
         Auth auth = Auth.getInstance();
         start(auth);
 
+        /* mock public client application*/
```
[the remark this answers is not in the document]
Comment (manuchandel) [reply sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java]: changed to caps

File: sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java (hunk 12 above)
Comment (manuchandel) [reply sdk/appcenter-auth/src/main/java/com/microsoft/appcenter/auth/Auth.java]: agreed. Thank you for explaining offline

File: sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java [hunk 14]
```diff
@@ -284,17 +293,45 @@ private void testInvalidConfig(JSONObject jsonConfig) throws Exception {
         Auth auth = Auth.getInstance();
         start(auth);
 
+        /* Mock public client application*/
```
Comment (guperrot) [inline sdk/appcenter-auth/src/test/java/com/microsoft/appcenter/auth/AuthTest.java]:
```suggestion
        /* Mock public client application. */
```

**Pull request merged**